### PR TITLE
[16.0][ADD] stock_intercompany_bidirectional

### PR DESCRIPTION
**Pull request by angelinaanaki**
This module extends stock_intercompany and allows to define both incoming and outgoing operation types. It allows to have a flexible bidirectional inventory flow between multiple companies in the system.

### CODE REVIEW

**Review by DemchukM — APPROVED** [review]
LGTM

File: stock_intercompany_bidirectional/models/res_company.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    intercompany_out_type_id = fields.Many2one(
+        "stock.picking.type", string="Intercompany Operation Type OUT"
+    )
+    intercompany_in_type_id = fields.Many2one(string="Intercompany Operation Type IN")
+    set_done_qty = fields.Boolean(string="Set Qty Done")
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/res_company.py]: It seems that the variable name `set_done_qty` could be more descriptive. A name that more clearly represents this action would significantly enhance code readability and maintainability. Maybe `auto_update_qty_done` with some additional help=`...`

File: stock_intercompany_bidirectional/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    move_packages = fields.Boolean()
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/res_company.py]: the same for `move_packages `

File: stock_intercompany_bidirectional/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        transactions depending on the operation types in the settings"""
+        active_company = self.env.company
+        inter_company = (
+            self.env["res.company"]
+            .sudo()
+            .search([("partner_id", "=", self.partner_id.id)], limit=1)
+        )
+
+        if (
+            active_company.intercompany_out_type_id.id == self.picking_type_id.id
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/stock_picking.py]: Just `self.env.company` a separate variable is redundant

File: stock_intercompany_bidirectional/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class StockPickingIntercompanyBidirectional(models.Model):
+    _inherit = "stock.picking"
+
+    def _create_counterpart_picking(self):
+        """Override of parent method to handle intercompany
+        transactions depending on the operation types in the settings"""
+        active_company = self.env.company
+        inter_company = (
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/stock_picking.py]: this is company related partner, maybe the `inter_company_partner` name is better

File: stock_intercompany_bidirectional/models/res_company.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2023 Cetmix OÜ
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    intercompany_out_type_id = fields.Many2one(
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/res_company.py:10]: intercompany_picking_type_out_id

File: stock_intercompany_bidirectional/models/res_company.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+        "stock.picking.type", string="Intercompany Operation Type OUT"
+    )
+    intercompany_in_type_id = fields.Many2one(string="Intercompany Operation Type IN")
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/res_company.py:13]: intercompany_picking_type_in_id

File: stock_intercompany_bidirectional/models/stock_picking.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        # Remove package-related data in move_line_ids
+        active_company = self.env.company
+        for line_data in move_line_ids:
+            line_data[2].update({"package_level_id": False, "package_id": False})
+
+        if company.set_done_qty:
+            # Update move_line_ids to include 'qty_done' based on counterpart move lines
+            for line_data in move_line_ids:
+                move_line = self.env["stock.move.line"].browse(
```
Comment (GabbasovDinar) [inline stock_intercompany_bidirectional/models/stock_picking.py]: I recommend optimizing the current loop by defining `stock_move_line_obj = self.env["stock.move.line"]` outside of the loop. In the current implementation, `self.env["stock.move.line"]` is called repeatedly in each iteration of the loop.

**Review by GabbasovDinar — CHANGES_REQUESTED** [review]
Thanks for the PR, Please, check my comments

File: stock_intercompany_bidirectional/models/res_company.py (hunk 5 above)
Comment (angelinaanaki) [reply stock_intercompany_bidirectional/models/res_company.py:10]: The name is defined in the 'stock_intercompany' (the module in depends), accessible in the [OCA/multi-company repository](https://github.com/OCA/multi-company/blob/16.0/stock_intercompany/models/res_company.py). Here, only the string for the field 'intercompany_in_type_id' has been modified, and a new field 'intercompany_out_type_id' has been introduced with the corresponding name. Considering these changes, do you recommend updating the names for better alignment and clarity?

**Comment by angelinaanaki** [comment]
@GabbasovDinar Thank you for your review. Please check and provide any feedback or recommendations you may have.

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
### Functional review

Tested on runboat, you can check there.

1. Create an operation of type "Delivery": Chicago -> San-Francisco [link](http://oca-multi-company-16-0-pr562-225d33be74d4.runboat.odoo-community.org/web#id=30&cids=4&menu_id=175&action=331&active_id=20&model=stock.picking&view_type=form) 

Put items in pack and validate the operation.
Validate it

2. Operation of type "Receipt" is created in the "San-Francisco" company [link](http://oca-multi-company-16-0-pr562-225d33be74d4.runboat.odoo-community.org/web#id=31&cids=1&menu_id=175&action=331&active_id=1&model=stock.picking&view_type=form)

**Expected behaviour**

- Receive From = Chicago
- Packages -> Visible in the package list
- After operation is validation packages must be in the location "WH/Stock"

**Current behaviour**

- Receive From = San-Francisco
- Not visible in the package list
- After operation is validation packages are in the location "Partner/Customers"


image

image

**Review by ivs-cetmix — CHANGES_REQUESTED** [review]
### Functional review

Tested on runboat

- Create an operation with product that has serial numbers enabled [link](http://oca-multi-company-16-0-pr562-546d9a74dae6.runboat.odoo-community.org/web#id=34&cids=1&menu_id=175&action=331&active_id=2&model=stock.picking&view_type=form)
- Confirm it. A counterpart picking is created in the other company
- Click the picking

**Expected behaviour**

Picking form view is opened

**Current behaviour**
Error
image

**Comment by angelinaanaki** [comment]
@ivs-cetmix Thank you for the review. Please check with the new configuration: "Mirror Lots/Serial Numbers".

**Review by ivs-cetmix — APPROVED** [review]
Functional review LGTM

File: stock_intercompany_bidirectional/models/res_company.py (hunk 5 above)
Comment (GabbasovDinar) [reply stock_intercompany_bidirectional/models/res_company.py:10]: no, then it makes sense to leave it as it is.

**Review by GabbasovDinar — APPROVED** [review]
Code LGTM

**Comment by ivs-cetmix** [comment]
Hey @OCA/intercompany-maintainers would appreciate someone checks this out. I think this module might be really useful in a lot of scenarios 😄

**Comment by sebastienbeau** [comment]
Hi, I see that there is already a similar work done here: https://github.com/OCA/multi-company/pull/520
It's was done directly in the stock_intercompany module.
@bguillot can you check the difference between the implementation ?

**Comment by ivs-cetmix** [comment]
Hi @sebastienbeau I've checked #520 and looks like module in the current PR has more features than that one. Anyway we are open to collaboration so let me know what should we do next.

**Review by GabbasovDinar — APPROVED** [review]

**Review by ivs-cetmix — APPROVED** [review]
Recent changes: functional review LGTM

**Comment by ivs-cetmix** [comment]
Hi @dreispt , we have updated the module and fixed some issues. Could you please check again?

**Pull request merged**